### ISSUE TITLE
Handling third-party libraries and getting version number from attr

### ISSUE THREAD

**Issue by timcu** [pypa/packaging-problems#546]
### OS version

Ubuntu 21.04

### Python version

Python 3.9.5

### Pip version

pip 21.2.4

### Guide link

https://packaging.python.org/tutorials/packaging-projects

### Problem description

Thank you for providing this packaging tutorial for simple projects. Unfortunately it is a little too simple and the guide at https://packaging.python.org/guides/distributing-packages-using-setuptools/ doesn't even mention `pyproject.toml` so is not much help. 

I have a simple project which is only different from you example project by three lines and one new file.

example_package/\_\_init.py\_\_
```
import toml
```
example_package/version.py
```
VERSION = "1.0.1"
```
setup.cfg (change line 3)
```
version = attr: example_package.version.VERSION
```
These would be standard requirements for a simple project
1. Use a third-party non-standard library
2. Keep version number in a single file from where it can be referenced

When I run `python3 -m build` I get the error message below. I can solve the error in one of two ways but neither seem correct. I would be interested in your advice.

Solution 1: Instead of `python3 -m build` use `python3 -m build --no-isolation`
Solution 2: Change pyproject.toml to read
```
[build-system]
requires = [
    "setuptools>=42",
    "wheel",
    "toml",
]
build-backend = "setuptools.build_meta"
```

### Error message

```text
(venv) tim@computer:~/PycharmProjects/packaging_tutorial$ python3 -m build
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools>=42, wheel)
* Getting dependencies for sdist...
Traceback (most recent call last):
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 419, in _parse_attr
    return getattr(StaticModule(module_name), attr_name)
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 26, in __init__
    spec = importlib.util.find_spec(name)
  File "/usr/lib/python3.9/importlib/util.py", line 94, in find_spec
    parent = __import__(parent_name, fromlist=['__path__'])
  File "/home/tim/PycharmProjects/packaging_tutorial/example_package/__init__.py", line 1, in <module>
    import toml
ModuleNotFoundError: No module named 'toml'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tim/PycharmProjects/packaging_tutorial/venv/lib/python3.9/site-packages/pep517/in_process/_in_process.py", line 349, in <module>
    main()
  File "/home/tim/PycharmProjects/packaging_tutorial/venv/lib/python3.9/site-packages/pep517/in_process/_in_process.py", line 331, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
  File "/home/tim/PycharmProjects/packaging_tutorial/venv/lib/python3.9/site-packages/pep517/in_process/_in_process.py", line 284, in get_requires_for_build_sdist
    return hook(config_settings)
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/build_meta.py", line 159, in get_requires_for_build_sdist
    return self._get_build_requires(config_settings, requirements=[])
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/build_meta.py", line 135, in _get_build_requires
    self.run_setup()
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/build_meta.py", line 150, in run_setup
    exec(compile(code, __file__, 'exec'), locals())
  File "setup.py", line 1, in <module>
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.9/distutils/core.py", line 121, in setup
    dist.parse_config_files()
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/dist.py", line 798, in parse_config_files
    parse_configuration(
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 158, in parse_configuration
    meta.parse()
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 498, in parse
    section_parser_method(section_options)
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 469, in parse_section
    self[name] = value
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 222, in __setitem__
    value = parser(value)
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 597, in _parse_version
    version = self._parse_attr(value, self.package_dir)
  File "/tmp/build-env-mtvhgcyo/lib/python3.9/site-packages/setuptools/config.py", line 422, in _parse_attr
    module = importlib.import_module(module_name)
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 855, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/tim/PycharmProjects/packaging_tutorial/example_package/__init__.py", line 1, in <module>
    import toml
ModuleNotFoundError: No module named 'toml'

ERROR Backend subproccess exited when trying to invoke get_requires_for_build_sdist
```

**Comment by KOLANICH** [comment]
just use `read_version`. `attr: ` relies on importing, and one mustn't import the lib itself into `setup.py`. Even better, use `setuptools_scm` instead.

**Comment by layday** [comment]
It's because `version` is a submodule, when importlib tries to find its "spec", as part of setuptool's locating the file to AST parse it, it will execute `__init__.py`.  You could try having the `VERSION` in `__init__.py` instead.

**Comment by timcu** [comment]
Thanks for these suggestions. Moving VERSION to `__init__.py` worked in the simple project described above but when I tried it in a more complex project I got a circular import.

I had a quick look into `read_version`. However I am a bit nervous using it because https://pypi.org/project/read-version/ says 

_Note: As of the release of setuptools 46.4.0, setuptools natively supports using the `attr:` directive in `setup.cfg` to set the project version based on static analysis of the project source. This makes the bulk of `read_version`’s use cases obsolete, and so it will not receive further development._

Next, I will see if I can use `setuptools_scm`
